---
title: "<em>IBDB</em> Documentation"
subtitle: "v1.0.0"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: "lumen"
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false 
    toc_depth: 2
    self_contained: false
params:
  object: "placeholder"
---

```{r setup, echo=FALSE, include=FALSE}
library(tidyverse)
library(shiny)
source("app/logic/d_documentation_utils.R")
```

![](logo.png){align="right" style="border: none; float: right;" width="360"}

**IBDB** is a web database for **accessing and exploring IBD expression data.**. The data for this database was derived from four studies profiled by *Ilieva et al., 2022*. **IBDB** is the work of the [Uchida laboratory](https://heartlncrna.github.io/){target="_blank"}, Center for RNA Medicine, Aalborg University, and the [Bioinformatics Research Network](http://www.bioresnet.org/){target="_blank"}.

Key features:

1. View transcriptomic data across two liver studies

2. Explore differential gene expression results

3. Download standardized and reprocessed datasets

## Datasets

The current version of **IBDB** contains the following datasets:

[1] GSE83687: A functional genomics predictive network model identifies regulators of inflammatory bowel disease: Mount Sinai Hospital (MSH) Population Specimen Collection and Profiling of Inflammatory Bowel Disease (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE83687)

[2] GSE112057: Whole Blood Transcriptome Profiling in Juvenile Idiopathic Arthritis and Inflammatory Bowel Disease (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112057)

[3] GSE123141: Intestinal macrophage mRNA sequencing of patients with colonic inflammatory bowel disease and a cohort of healthy controls (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123141)

## IBDB Interfaces -- **UNFINISHED**

![**The 'Results' page interface**. (**A**) Controls the study which is displayed. (**B**) Controls which contrast is displayed for plots that subsets by contrast. (**C**) The Results table which displays the results of differential gene expression analysis for each gene in the study. Selecting a row will change the output in **E**. (**D**) Selector which controls the type of expression data displayed in **E**. (**E**) Interactive box plots showing the distribution of gene expression values across samples within each biological condition in the selected study.](usage_liverdb_1.png){align="right" style="border: none; float: none;" width="1000"}

![**DGE volcano plots**. This interface displays the results of DGE analysis as a volcano plot (highlighted). Selecting a row in the 'Results Table' will cause the corresponding gene to be highlighted in the volcano plot.](usage_liverdb_2.png){align="right" style="border: none; float: none;" width="1000"}

![**DGE Heatmap**. This interface displays the top hits from DGE analysis as a heatmap for the selected study.](usage_liverdb_3.png){align="right" style="border: none; float: none;" width="1000"}

![**Pathway enrichment plot**. This interface displays the top results from KEGG pathway analysis as a heatmap. A drop down is available for controlling the displayed enrichment metric. "Combined.Score" - a combination of the p value from Fisher's exact test and z score from expected enrichment; "Odds.Ratio" - odds ratio from Fisher's exist test; "Padj (-log10)" - Adjusted p value from enrichment analysis, log transformed. For more info, see the [enrichr](https://maayanlab.cloud/Enrichr/) documentation.](usage_liverdb_4.png){align="right" style="border: none; float: none;" width="1000"}

![**DGE UpSet Plot**. This interface displays an UpSet plot which is used to determine the number of DEGs seen mutually across study/contrast pairs. A drop down menu is used to control whether Over- or Under-expressed DEGs are compared.](usage_liverdb_5.png){align="right" style="border: none; float: none;" width="1000"}

## Terminology

- *DGE*: Differential Gene Expression
  - Differential Gene Expression refers to a significant difference in the expression of a gene between two conditions of interest. In this study, we calculated DGE using the [egdeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html){target="_blank"} R/Bioconductor package.
- *CPM*: Counts per Million
  - Read counts normalized by library size. In this case, they were normalized using the moderated counts-per-million method in [egdeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html){target="_blank"}.
- *TPM*: Transcripts per Million
  - Read counts normalized based on transcript length and read depth (different from RPKM). [More info](https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/){target="_blank"}.
- *RPKM*: Reads Per Kilobase of transcript, per Million mapped reads
  - Read counts normalized based on transcript length and read depth. [More info](https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/){target="_blank"}.


## Bugs

IBDB is a new database, which means bugs may occasionally occur. If you encounter any bugs or unexpected behavior, please [open an issue](https://github.com/RVel37/IBDB/issues/) on the GitHub repo and describe, in as much detail as possible, the following:

1.  What you expected IBDB to do.
2.  What IBDB did and why it was unexpected.
3.  Any error messages you received (along with screenshots).

## License and attribution

IBDB is licensed under an MIT license and we ask that you please cite IBDB in any published work like so:

> Ilieva et al., LiverDB: Systematic analysis of long non-coding RNA genes
in nonalcoholic fatty liver disease, 2022
